### PR TITLE
gh-95853: Improve WASM build script (GH-96389)

### PR DESCRIPTION
**Pull request by tiran**
- pre-build Emscripten ports and system libraries
- check for broken EMSDK versions
- use EMSDK's node for wasm32-emscripten
- warn when PKG_CONFIG_PATH is set
- add support level information

* Issue: gh-95853

### CODE REVIEW

**Review by rchiodo — COMMENTED** [review]
Thanks for this

File: Tools/wasm/wasm_build.py [hunk 1]
```diff
@@ -80,24 +87,28 @@
 """
 
 
-def get_emscripten_root(emconfig: pathlib.Path = EM_CONFIG) -> pathlib.PurePath:
-    """Parse EM_CONFIG file and lookup EMSCRIPTEN_ROOT
+def get_emscripten_root(
+    emconfig: pathlib.Path = EM_CONFIG,
+) -> Iterable[pathlib.PurePath]:
+    """Parse EM_CONFIG file and lookup EMSCRIPTEN_ROOT and NODE_JS
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
    """Parse EM_CONFIG file and lookup EMSCRIPTEN_ROOT and NODE_JS.
```

File: Tools/wasm/wasm_build.py [hunk 2]
```diff
@@ -80,24 +87,28 @@
...
 
     The ".emscripten" config file is a Python snippet that uses "EM_CONFIG"
     environment variable. EMSCRIPTEN_ROOT is the "upstream/emscripten"
     subdirectory with tools like "emconfigure".
     """
     if not emconfig.exists():
-        return _MISSING
+        return _MISSING, _MISSING
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py:100]: Why the double return? Is this just to make it an iterable, or specifically an iterable of 2? If the latter then:

```suggestion
        return [_MISSING]
```

If the latter then should this function be changed to return a tuple of two paths?

File: Tools/wasm/wasm_build.py [hunk 3]
```diff
@@ -153,6 +164,9 @@ class Platform:
     make_wrapper: Optional[pathlib.PurePath]
     environ: dict
     check: Callable[[], None]
+    # used for build_emports()
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
    # Used for build_emports().
```

File: Tools/wasm/wasm_build.py [hunk 4]
```diff
@@ -198,12 +214,24 @@ def _check_emscripten():
...
             f"Emscripten SDK {version} in '{EMSCRIPTEN_ROOT}' is older than "
             "minimum required version "
             f"{'.'.join(str(v) for v in EMSDK_MIN_VERSION)}.",
         )
+    broken = EMSDK_BROKEN_VERSION.get(version_tuple)
+    if broken is not None:
+        raise ConditionError(
+            os.fspath(version_txt),
+            f"Emscripten SDK {version} in '{EMSCRIPTEN_ROOT}' has known ",
+            f"bugs, see {broken}.",
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]: I think Black inserted a comma you didn't want.

```suggestion
            (f"Emscripten SDK {version} in '{EMSCRIPTEN_ROOT}' has known "
            f"bugs, see {broken}."),
```

File: Tools/wasm/wasm_build.py [hunk 5]
```diff
@@ -280,6 +314,42 @@ def is_wasi(self) -> bool:
         cls = type(self)
         return self in {cls.wasm32_wasi, cls.wasm64_wasi}
 
+    def get_extra_paths(self) -> Iterable[pathlib.PurePath]:
+        """Host-specific os.environ["PATH"] entries
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
        """Host-specific os.environ["PATH"] entries.
```

File: Tools/wasm/wasm_build.py [hunk 6]
```diff
@@ -280,6 +314,42 @@ def is_wasi(self) -> bool:
...
+            return [NODE_JS.parent]
+        elif self == cls.wasm64_emscripten:
+            # TODO: look for recent node
+            return []
+        else:
+            return []
+
+    @property
+    def emport_args(self) -> List[str]:
+        """Host-specific port args (Emscripten)"""
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
        """Host-specific port args (Emscripten)."""
```

File: Tools/wasm/wasm_build.py [hunk 7]
```diff
@@ -280,6 +314,42 @@ def is_wasi(self) -> bool:
...
+        if self is cls.wasm64_emscripten:
+            return ["-sMEMORY64=1"]
+        elif self is cls.wasm32_emscripten:
+            return ["-sMEMORY64=0"]
+        else:
+            return []
+
+    @property
+    def embuilder_args(self) -> List[str]:
+        """Host-specific embuilder args (Emscripten)"""
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
        """Host-specific embuilder args (Emscripten)."""
```

File: Tools/wasm/wasm_build.py [hunk 8]
```diff
@@ -294,10 +364,32 @@ def can_execute(self) -> bool:
         cls = type(self)
         return self not in {cls.browser, cls.browser_debug}
 
+    @property
+    def emport_args(self) -> List[str]:
+        """Target-specific port args"""
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
        """Target-specific port args."""
```

File: Tools/wasm/wasm_build.py [hunk 9]
```diff
@@ -380,6 +472,12 @@ def getenv(self) -> dict:
         for key, value in platenv.items():
             if value is None:
                 env.pop(key, None)
+            elif key == "PATH":
+                # list of path items, prefix with extra paths
+                new_path: List[pathlib.PurePath] = []
+                new_path.extend(self.host.get_extra_paths())
+                new_path.extend(value)
+                env[key] = os.pathsep.join(os.fspath(p) for p in new_path)
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py:482]: Just a suggestion; in no way important. 😄

```suggestion
                env[key] = os.pathsep.join(map(os.fspath, new_path))
```

File: Tools/wasm/wasm_build.py [hunk 10]
```diff
@@ -443,10 +548,57 @@ def clean(self, all: bool = False):
         elif self.makefile.exists():
             self.run_make("clean")
 
+    def build_emports(self, force: bool = False):
+        """Pre-build emscripten ports"""
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
        """Pre-build emscripten ports."""
```

File: Tools/wasm/wasm_build.py [hunk 11]
```diff
@@ -443,10 +548,57 @@ def clean(self, all: bool = False):
...
+        if force:
+            embuilder_cmd.append("--force")
+
+        ports_cmd = [os.fspath(platform.cc)]
+        ports_cmd.extend(self.host.emport_args)
+        if self.target:
+            ports_cmd.extend(self.target.emport_args)
+
+        if self.dynamic_linking:
+            # trigger PIC build
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
            # Trigger PIC build.
```

File: Tools/wasm/wasm_build.py [hunk 12]
```diff
@@ -443,10 +548,57 @@ def clean(self, all: bool = False):
...
+            ports_cmd.append("-sMAIN_MODULE")
+            embuilder_cmd.append("--pic")
+        if self.pthreads:
+            # trigger multi-threaded build
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
            # Trigger multi-threaded build.
```

File: Tools/wasm/wasm_build.py [hunk 13]
```diff
@@ -443,10 +548,57 @@ def clean(self, all: bool = False):
...
+            ports_cmd.append("-sUSE_PTHREADS")
+            # embuilder_cmd.append("--pthreads")
+
+        # pre-build libbz2, libsqlite3, libz, and some system libs
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py]:
```suggestion
        # Pre-build libbz2, libsqlite3, libz, and some system libs.
```

File: Tools/wasm/wasm_build.py [hunk 14]
```diff
@@ -443,10 +548,57 @@ def clean(self, all: bool = False):
...
+            ports_cmd.extend(self.target.emport_args)
+
+        if self.dynamic_linking:
+            # trigger PIC build
+            ports_cmd.append("-sMAIN_MODULE")
+            embuilder_cmd.append("--pic")
+        if self.pthreads:
+            # trigger multi-threaded build
+            ports_cmd.append("-sUSE_PTHREADS")
+            # embuilder_cmd.append("--pthreads")
```
Comment (brettcannon) [inline Tools/wasm/wasm_build.py:577]: Should this go? Or is this here for when it will be useful?
```suggestion
```

**Review by brettcannon — APPROVED** [review]

File: Tools/wasm/wasm_build.py (hunk 14 above)
Comment (tiran) [reply Tools/wasm/wasm_build.py:577]: It's a reminder. I have a PR for next Emscripten version that adds the option.

File: Tools/wasm/wasm_build.py (hunk 2 above)
Comment (tiran) [reply Tools/wasm/wasm_build.py:100]: It's me getting the hang out of typing. You are right, a structured tuple is the better return type here.

**Pull request merged**